scripts/check-title-labels.lean

```diff
@@ -0,0 +1,267 @@
+Authors: Michael Rothgang
+-/
+
+import Cli.Basic
+import Batteries.Data.HashMap
+import Batteries.Data.String.Matcher
+
+/-!
+# Checker for well-formed title and labels
+
+This script checks if a PR title matches mathlib's commit conventions,
+and if the PR has any contradictory labels.
+Not all checks from the commit conventions are implemented: for instance, no effort is made to
+verify whether the title or body are written in present imperative tense.
+-/
+
+/-- Split a string in two, at a given position. Assumes `pos` is valid. -/
+-- FIXME: this should be added to batteries, as a version of "splitOn"
+def splitAtPos (s : String) (pos : String.Pos) : String × String :=
+  let before := s.extract 0 pos
+  let after := s.extract (s.next pos) s.endPos
+  (before, after)
+
+/--
+Check if `title` matches the mathlib conventions for PR titles
+(documented at https://leanprover-community.github.io/contribute/commit.html).
+Not all checks are implemented: for instance, no effort is made to verify if the title or body
+are written in present imperative tense.
+
+Return all error messages for violations found.
+-/
+def validateTitle (title: String) : Array String := Id.run do
+  -- The title should be of the form "abbrev: main title" or "abbrev(scope): main title".
+  if !title.contains ':' then
+    return #["error: the PR title does not contain a colon"]
+
+  -- Split 'title' at the first occurrence of ':'.
+  let (type, main) := splitAtPos title (title.find (· == ':'))
+  let mut errors := #[]
+  if main.endsWith "." then
+    errors := errors.push "error: the PR title should not end with a full stop"
+
+  if main.front != ' ' then
+    errors := errors.push "error: the PR should have the form 'abbrev: main title', with a space"
+  else if main.containsSubstr "  " then
+    errors := errors.push "error: the PR title contains multiple consecutive spaces; please add just one"
+  let main := main.removeLeadingSpaces
+  if main.front.toLower != main.front then
+    errors := errors.push "error: the main PR title should be lowercased"
+
+  -- `type` should be of the form abbrev or abbrev(scope), where `scope` is of the form
+  -- `Algebra/Basic`, without a leading `Mathlib` or a trailing `.lean`
+  let N := (type.splitOn "(").length
+  let M := (type.splitOn ")").length
+  let known_types := ["feat", "chore", "perf", "refactor", "style", "fix", "doc", "test"]
+  if (N, M) == (1, 1) then
+    if !known_types.contains type then
+      errors := errors.push s!"error: the PR type should be one of {", ".intercalate (known_types.map (s!"\"{·}\""))}"
+  else if (N, M) == (2, 2) then
+    if !type.endsWith ")" then
+      errors := errors.push s!"error: the PR type should be of the form abbrev(scope)"
+    let idx := type.find (· == '(')
+    let idx2 := type.find (· == ')')
+    if idx2 < idx then
+      errors := errors.push "error: mismatched parentheses; the PR type should be of the form abbrev(scope)"
```

Comment:
This should be covered by checking that the type ends with a `)`

Reply:
It mostly was, but not quite. I've added a few more tests --- but thanks for pointing me in this direction!